src/scan.c

```diff
@@ -116,11 +117,19 @@ static int scan_module(struct context_data *ctx, int ep, int chain)
     ord2 = -1;
     ord = ep - 1;
 
-    gvol_memory = break_row = row_count = frame_count = 0;
+    gvol_memory = break_row = row_count = row_count_total = frame_count = 0;
+    orders_since_last_valid = any_valid = 0;
     start_time = time = 0.0;
     inside_loop = 0;
 
     while (42) {
+	/* Sanity check to prevent getting stuck due to broken patterns. */
+	if (orders_since_last_valid > 512) {
+	    D_(D_CRIT "orders_since_last_valid = %d @ ord %d; ending scan", orders_since_last_valid, ord);
```

Comment:
Should this imply a failure?  I.e. should we set `any_valid` to 0 here?

Reply:
I don't think so—some valid tracks (particularly ones containing a lot of loops) are capable of reaching this. This OctaMED module should loop for something like an hour(!) and encounters this check, but it is a valid module.

https://modland.com/pub/modules/OctaMED%20MMD1/Apollon/pilot%20i%20tyskland.mmd1

Reply:
Sorry, I deleted a comment I made thinking this was the other >512 check. I think it theoretically might be possible to cause this check to occur and still have a partial valid module. I'm not entirely sure.

Reply:
Will you investigate more, or should we merge?

Reply:
I tried yesterday to create a test case that could cause that state to occur but failed to, so I think I'll probably have to step through the code. You made a valid point and what I suspect could occur is that if this check is flagged but allows the scan to succeed (`any_valid`=1), the player scan could later get stuck in an infinite loop (a problem I encountered before adding `any_valid` at all). So the solution may be to remove `any_valid` and just unconditionally return -1 here.

Reply:
Okay, after some experimentation, I think that the current implementation should be fine—as long as the sequence has a valid entry point it won't get stuck in an infinite loop in `next_order` and `any_valid` makes sure that the entry point (eventually) reaches something valid.

Reply:
OK then.